system/jlib/jutil.cpp

```diff
@@ -3056,19 +3175,481 @@ extern jlib_decl void setSecretMount(const char * path)
+            return false;
+        return it->second->requestSecret(kind, content, secret, version);
+    }
+};
+
+class CVaultManager : public CInterfaceOf<IVaultManager>
+{
+private:
+    std::map<std::string, std::unique_ptr<CVaultSet>> categories;
+public:
+    CVaultManager()
+    {
+        IPropertyTree *config = queryComponentConfig().queryPropTree("vaults");
+        if (!config)
+            return;
+        Owned<IPropertyTreeIterator> iter = config->getElements("*");
+        ForEach (*iter)
+        {
+            IPropertyTree &cat = iter->query();
+            const char *name = cat.queryName();
+            categories.emplace(name, std::unique_ptr<CVaultSet>(new CVaultSet(&cat)));
+        }
+    }
+    bool getCachedSecretFromVault(const char *category, const char *vaultId, CVaultKind &kind, StringBuffer &content, const char *secret, const char *version) override
+    {
+        if (isEmptyString(category))
+            return false;
+        std::map<std::string, std::unique_ptr<CVaultSet>>::iterator it = categories.find(category);
+        if (it == categories.end())
+            return false;
+        return it->second->getCachedSecretFromVault(vaultId, kind, content, secret, version);
+    }
+    bool requestSecretFromVault(const char *category, const char *vaultId, CVaultKind &kind, StringBuffer &content, const char *secret, const char *version) override
+    {
+        if (isEmptyString(category))
+            return false;
+        std::map<std::string, std::unique_ptr<CVaultSet>>::iterator it = categories.find(category);
+        if (it == categories.end())
+            return false;
+        return it->second->requestSecretFromVault(vaultId, kind, content, secret, version);
+    }
+
+    bool getCachedSecretByCategory(const char *category, CVaultKind &kind, StringBuffer &content, const char *secret, const char *version) override
+    {
+        if (isEmptyString(category))
+            return false;
+        std::map<std::string, std::unique_ptr<CVaultSet>>::iterator it = categories.find(category);
+        if (it == categories.end())
+            return false;
+        return it->second->getCachedSecret(kind, content, secret, version);
+    }
+    bool requestSecretByCategory(const char *category, CVaultKind &kind, StringBuffer &content, const char *secret, const char *version) override
+    {
+        if (isEmptyString(category))
+            return false;
+        std::map<std::string, std::unique_ptr<CVaultSet>>::iterator it = categories.find(category);
+        if (it == categories.end())
+            return false;
+        return it->second->requestSecret(kind, content, secret, version);
+    }
+};
+
+static Owned<CVaultManager> vaultManager;
+IVaultManager *ensureVaultManager()
+{
+    CriticalBlock block(secretCS);
+    if (!vaultManager)
+        vaultManager.setown(new CVaultManager());
+    return vaultManager;
+}
+
+CriticalSection secretCacheCS;
+static Owned<IPropertyTree> secretCache = createPTree();
```

Comment:
need to be careful about construction ordering.  I'm not sure how much of a problem it is these days, but safer to use a MODULE_INIT to initialize because jptree has some interesting initialization.